src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -1383,9 +1383,29 @@ private void CheckAllowedProjectChanges(ProjectChanges projectChanges)
             }
         }
 
+        /// <summary>
+        /// Called during a call to <see cref="TryApplyChanges(Solution)"/> to determine if a specific change to <see cref="Project.CompilationOptions"/> is allowed.
+        /// </summary>
+        /// <remarks>
+        /// This method is only called if <see cref="CanApplyChange" /> returns false for <see cref="ApplyChangesKind.ChangeCompilationOptions"/>.
+        /// If <see cref="CanApplyChange" /> returns true, then that means all changes are allowed and this method does not need to be called.
+        /// </remarks>
+        /// <param name="oldOptions">The old <see cref="CompilationOptions"/> of the project from prior to the change.</param>
+        /// <param name="newOptions">The new <see cref="CompilationOptions"/> of the project that was passed to <see cref="TryApplyChanges(Solution)"/>.</param>
+        /// <param name="project">The project contained in the <see cref="Solution"/> passed to <see cref="TryApplyChanges(Solution)"/>.</param>
```

Comment:
@ryzngard Does this confirm your understanding of our contract? At first I admit I found this a bit surprising, but I guess that's where the long conversation ended in https://github.com/dotnet/roslyn/pull/25875#discussion_r179558610?

Reply:
(also tagging @jcouv and @Neme12 who were in that long thread too...)

Reply:
How I used it in https://github.com/dotnet/roslyn/blob/master/src/Features/CSharp/Portable/UpgradeProject/CSharpUpgradeProjectCodeFixProvider.cs#L103 isn't exactly this way, since I didn't check `CanApplyChange` first.

Reply:
Ah, hmm. Should we have the default implementation of CanApplyCompilationOptionChange just call CanApplyChange(ApplyChangeKind.CompilationOptions) then, so anybody who didn't override this helper falls back?

Reply:
That seems reasonable to me, since we expect the overhead to be very small no harm in calling it twice if they do the right thing.

Reply:
Well, if somebody did implement the right pattern, it's an or boolean operation anyways: once you got a true from one function, you shouldn't be checking the other no matter what.